tests/testsuite/publish.rs

```diff
@@ -2048,3 +2049,187 @@ error: package ID specification `bar` did not match any packages
         )
         .run();
 }
+
+#[cargo_test]
+fn wait_for_publish() {
+    // Counter for number of tries before the package is "published"
+    let arc: Arc<Mutex<u32>> = Arc::new(Mutex::new(0));
+    let arc2 = arc.clone();
+
+    // Registry returns an invalid response.
+    let _registry = registry::RegistryBuilder::new()
+        .http_index()
+        .http_api()
+        .add_responder("/index/de/la/delay", move |req, server| {
+            let mut lock = arc.lock().unwrap();
+            *lock += 1;
+            // if the package name contains _ or
+            if *lock <= 1 {
+                server.not_found(req)
+            } else {
+                server.index(req)
+            }
+        })
+        .build();
+
+    // The sparse-registry test server does not know how to publish on its own.
+    // So let us call publish for it.
+    Package::new("delay", "0.0.1")
+        .file("src/lib.rs", "")
+        .publish();
+
+    let p = project()
+        .file(
+            "Cargo.toml",
+            r#"
+                [project]
+                name = "delay"
+                version = "0.0.1"
+                authors = []
+                license = "MIT"
+                description = "foo"
+
+            "#,
+        )
+        .file("src/lib.rs", "")
+        .build();
+
+    p.cargo("publish --no-verify -Z sparse-registry")
+        .masquerade_as_nightly_cargo(&["sparse-registry"])
+        .with_status(0)
+        .with_stderr(
+            "\
+[UPDATING] `dummy-registry` index
+[WARNING] using `registry.token` config value with source replacement is deprecated
+This may become a hard error in the future; see <https://github.com/rust-lang/cargo/issues/xxx>.
+Use the --token command-line flag to remove this warning.
+[WARNING] manifest has no documentation, [..]
+See [..]
+[PACKAGING] delay v0.0.1 ([CWD])
+[UPLOADING] delay v0.0.1 ([CWD])
+[UPDATING] `dummy-registry` index
+[WAITING] on registry `dummy-registry` to finish publish
+[UPDATING] `dummy-registry` index
```

Comment:
Since this isn't ideal, I wanted to call out that we'll post an "Updating" message each time we poll.

This is because we are going through the registry source path instead of the registry API path in hopes to wait out any CDN propagation delays.

I'd like to silence these but can't think of any good option.  I could temporarily make the `Shell` "quiet" and then re-enable the previous verbosity level but the risks with that make me unsure if its worth it.

(see my other post about the "Waiting" message wording)

Reply:
Now that we are polling on the `Source` rather than `PackageRegistry`, this has gone away